File: src/main/java/rars/exceptions/ExceptionReason.java
package rars.exceptions;

public enum ExceptionReason {
    SOFTWARE_INTERRUPT(0x80000000),
    TIMER_INTERRUPT(0x80000004),
    EXTERNAL_INTERRUPT(0x80000008),
    INSTRUCTION_ADDR_MISALIGNED(0),
    INSTRUCTION_ACCESS_FAULT(1),
    ILLEGAL_INSTRUCTION(2),
    LOAD_ADDRESS_MISALIGNED(4),
    LOAD_ACCESS_FAULT(5),
    STORE_ADDRESS_MISALIGNED(6),
    STORE_ACCESS_FAULT(7),
    ENVIRONMENT_CALL(8),
    OTHER(-1);
    public final int value;

    ExceptionReason(final int value) {
        this.value = value;
    }

    public boolean isInterrupt() {
        return switch (this) {
            case SOFTWARE_INTERRUPT, TIMER_INTERRUPT, EXTERNAL_INTERRUPT ->
                    true;
            default -> false;
        };
    }
}

File: rars.runtime/src/commonMain/kotlin/rars/runtime/ExceptionReason.kt
package rars.runtime

sealed interface ExceptionReason {
    val value: Int

    data object IllegalInstruction : ExceptionReason {
        override val value = 0x2
    }

    data object EnvironmentCall : ExceptionReason {
        override val value = 0x8
    }

    data object Other : ExceptionReason {
        override val value = -0x1
    }

    enum class Interrupt(override val value: Int) : ExceptionReason {
        Software(-0x80000000),
        Timer(-0x7ffffffc),
        External(-0x7ffffff8),
    }

    enum class Memory(override val value: Int) : ExceptionReason {
        InstructionAddressMisaligned(0x0),
        InstructionAddressFault(0x1),
        LoadAddressMisaligned(0x4),
        LoadAddressFault(0x5),
        StoreAddressMisaligned(0x6),
        StoreAddressFault(0x7);

    }
}